### PR TITLE
Make DepGraph::previous_work_products immutable

### LINKED ISSUE #50501: [parallel-queries] DepGraph::previous_work_products could be made immutable to avoid shared mutable state
[DepGraph::previous_work_products](https://github.com/rust-lang/rust/blob/c88992d00169f15d8ff9edac8901245b4d21e1c1/src/librustc/dep_graph/graph.rs#L76-L80) is a field in the dep-graph that contains the list of object files cached during the previous compilation session. In theory, this information is available already at the time when we load the previous dep-graph and thus the field could be made immutable (i.e. not use an `RwLock`) and be initialized at `DepGraph` construction time.

Refactoring this would need the following steps:
- [Load the previous work products data](https://github.com/rust-lang/rust/blob/c88992d00169f15d8ff9edac8901245b4d21e1c1/src/librustc_incremental/persist/load.rs#L43-L78) already in [load_dep_graph()](https://github.com/rust-lang/rust/blob/c88992d00169f15d8ff9edac8901245b4d21e1c1/src/librustc_incremental/persist/load.rs#L151) instead of in [dep_graph_tcx_init()](https://github.com/rust-lang/rust/blob/c88992d00169f15d8ff9edac8901245b4d21e1c1/src/librustc_incremental/persist/load.rs#L28)
- Pass previous work products data as parameter to [DepGraph::new()](https://github.com/rust-lang/rust/blob/c88992d00169f15d8ff9edac8901245b4d21e1c1/src/librustc/dep_graph/graph.rs#L93)
- Remove the `RwLock` around the [DepGraph::previous_work_products](https://github.com/rust-lang/rust/blob/c88992d00169f15d8ff9edac8901245b4d21e1c1/src/librustc/dep_graph/graph.rs#L76-L80) field.

cc @rust-lang/wg-compiler-performance

### PR DESCRIPTION
**Pull request by wesleywiser**
Fixes #50501 

r? @michaelwoerister

### CODE REVIEW

File: src/librustc_incremental/persist/load.rs [hunk 1]
```diff
@@ -169,6 +127,50 @@ pub fn load_dep_graph(sess: &Session) ->
     let report_incremental_info = sess.opts.debugging_opts.incremental_info;
     let expected_hash = sess.opts.dep_tracking_hash();
 
+    let mut prev_work_products = FxHashMap();
+
+    // If we are only building with -Zquery-dep-graph but without an actual
+    // incr. comp. session directory, we exit here. Otherwise we'd fail
+    // when trying to load work products.
```
Comment (michaelwoerister) [inline src/librustc_incremental/persist/load.rs]: I'd reword this comment since we don't really exit the function if there is no cache directory. It could just say something like "If there is an incr. comp. cache directory, load the set of previous work products".

**Comment by michaelwoerister** [comment]
Excellent, thanks a lot, @wesleywiser! Please update that one comment, then this is good to go.

**Comment by wesleywiser** [comment]
Thanks @michaelwoerister! I updated that comment.

**Comment by michaelwoerister** [comment]
Thanks!

@bors r+
